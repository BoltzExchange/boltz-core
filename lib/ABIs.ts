import ERC20ABI from '../out/ERC20.sol/ERC20.json';
import ERC20SwapABI from '../out/ERC20Swap.sol/ERC20Swap.json';
import EtherSwapABI from '../out/EtherSwap.sol/EtherSwap.json';
import RouterABI from '../out/Router.sol/Router.json';

export const ContractABIs = {
  ERC20: ERC20ABI.abi,
  ERC20Swap: ERC20SwapABI.abi,
  EtherSwap: EtherSwapABI.abi,
  Router: RouterABI.abi,
} as const;
