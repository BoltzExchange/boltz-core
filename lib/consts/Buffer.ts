export const Nonce = Buffer.from('00', 'hex');
export const EmptyScript = Buffer.alloc(0);
export const PrefixUnconfidential = Buffer.from('01', 'hex');
