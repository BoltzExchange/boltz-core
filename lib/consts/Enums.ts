export enum OutputType {
  Bech32 = 0,
  Compatibility = 1,
  Legacy = 2,
  Taproot = 3,
}
