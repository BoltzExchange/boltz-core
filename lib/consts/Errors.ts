const Errors = {
  TAPROOT_NOT_SUPPORTED: 'taproot not supported',
};

export default Errors;
