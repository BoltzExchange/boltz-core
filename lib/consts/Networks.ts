import { networks } from 'bitcoinjs-lib';

const bitcoinPrefix = '\\x18Bitcoin Signed Message:\n';
const litecoinPrefix = '\\x19Litecoin Signed Message:\n';
const dogecoinPrefix = '\\x19Dogecoin Signed Message:\n';

const testnetBip32 = {
  public: 0x019DA462,
  private: 0x019D9CFE,
};

const simnetBip32 = {
  public: 0x420BD3A,
  private: 0x488B21E,
};

const Networks = {
  // Bitcoin networks
  bitcoinMainnet: networks.bitcoin,
  bitcoinTestnet: networks.testnet,
  bitcoinSimnet: {
    messagePrefix: bitcoinPrefix,
    bip32: simnetBip32,
    bech32: 'sb',
    scriptHash: 0x7b,
    pubKeyHash: 0x3f,
    wif: 0x64,
  },
  bitcoinRegtest: networks.regtest,

  // Litecoin networks
  litecoinMainnet: {
    messagePrefix: litecoinPrefix,
    bip32: {
      private: 0x488ADE4,
      public: 0x488B21E,
    },
    bech32: 'ltc',
    scriptHash: 0x32,
    pubKeyHash: 0x30,
    wif: 0xb0,
  },
  litecoinTestnet: {
    messagePrefix: litecoinPrefix,
    bip32: testnetBip32,
    bech32: 'tltc',
    scriptHash: 0x3a,
    pubKeyHash: 0x6f,
    wif: 0xef,
  },
  litecoinSimnet: {
    messagePrefix: bitcoinPrefix,
    bip32: simnetBip32,
    bech32: 'sltc',
    scriptHash: 0x7b,
    pubKeyHash: 0x3f,
    wif: 0x64,
  },
  litecoinRegtest: {
    messagePrefix: litecoinPrefix,
    bip32: testnetBip32,
    bech32: 'rltc',
    scriptHash: 0x3a,
    pubKeyHash: 0x6f,
    wif: 0xef,
  },

  // Dogecoin networks
  dogecoinMainnet: {
    messagePrefix: dogecoinPrefix,
    bip32: {
      public: 0x02facafd,
      private: 0x02fac398,
    },
    pubKeyHash: 0x1e,
    scriptHash: 0x16,
    wif: 0x9e,
  },
  dogecoinTestnet: {
    messagePrefix: dogecoinPrefix,
    bip32: testnetBip32,
    pubKeyHash: 0x71,
    scriptHash: 0xc4,
    wif: 0xf1,
  },
  dogecoinRegtest: {
    messagePrefix: dogecoinPrefix,
    bip32: testnetBip32,
    pubKeyHash: 0x6f,
    scriptHash: 0xc4,
    wif: 0xef,
  },
};

export default Networks;
