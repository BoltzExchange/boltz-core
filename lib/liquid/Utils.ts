import { confidential, TxOutput } from 'liquidjs-lib';
import { confidentialLiquid } from './Init';

export const getOutputValue = (
  output: TxOutput & {
    blindingPrivKey?: Buffer;
  },
): number => {
  return output.blindingPrivKey
    ? Number(
        confidentialLiquid.unblindOutputWithKey(output, output.blindingPrivKey)
          .value,
      )
    : confidential.confidentialValueToSatoshi(output.value);
};
