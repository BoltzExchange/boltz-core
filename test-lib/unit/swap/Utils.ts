import Errors from '../../../lib/consts/Errors';
import { OutputType } from '../../../lib/consts/Enums';
import { p2wshOutput, p2shOutput, p2shP2wshOutput } from '../../../lib/swap/Scripts';

export const getScriptHashFunction = (type: OutputType): (scriptHex: Buffer) => Buffer => {
  switch (type) {
    case OutputType.Bech32: return p2wshOutput;
    case OutputType.Legacy: return p2shOutput;
    case OutputType.Compatibility: return p2shP2wshOutput;
    case OutputType.Taproot: throw Errors.TAPROOT_NOT_SUPPORTED;
  }
};
