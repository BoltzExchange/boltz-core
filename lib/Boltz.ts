import Networks from './consts/Networks';
import * as Scripts from './swap/Scripts';
import { OutputType } from './consts/Enums';
import * as SwapUtils from './swap/SwapUtils';
import { swapScript } from './swap/SwapScript';
import { detectSwap } from './swap/SwapDetector';
import { constructClaimTransaction } from './swap/Claim';
import { detectPreimage } from './swap/PreimageDetector';
import { constructRefundTransaction } from './swap/Refund';
import { estimateFee, estimateSize } from './FeeCalculator';
import { ScriptElement, TransactionOutput } from './consts/Types';

export {
  Networks,
  OutputType,
  ScriptElement,
  TransactionOutput,

  Scripts,

  swapScript,

  detectSwap,
  detectPreimage,

  estimateFee,
  estimateSize,

  constructClaimTransaction,
  constructRefundTransaction,

  SwapUtils,
};
