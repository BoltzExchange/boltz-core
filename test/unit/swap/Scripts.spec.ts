import {
  outputFunctionForType,
  p2pkhOutput,
  p2shOutput,
  p2shP2wpkhOutput,
  p2shP2wshOutput,
  p2wpkhOutput,
  p2wshOutput,
} from '../../../lib/swap/Scripts';
import { OutputType } from '../../../lib/consts/Enums';
import { getHexBuffer, getHexString } from '../../../lib/Utils';

describe('Scripts', () => {
  const publicKeyHash = getHexBuffer(
    '0000000000000000000000000000000000000000',
  );
  const redeemScript = getHexBuffer('00');

  test.each`
    scriptFunc         | input            | name
    ${p2wpkhOutput}    | ${publicKeyHash} | ${'P2WPKH'}
    ${p2wshOutput}     | ${redeemScript}  | ${'P2WSH'}
    ${p2pkhOutput}     | ${publicKeyHash} | ${'P2PKH'}
    ${p2shOutput}      | ${redeemScript}  | ${'P2SH'}
    ${p2shP2wshOutput} | ${redeemScript}  | ${'P2SH nested P2WSH'}
  `(`should get $name output script`, async ({ scriptFunc, input }) => {
    expect(getHexString(scriptFunc(input))).toMatchSnapshot();
  });

  test('should get P2SH nested P2WPKH output script', () => {
    expect(p2shP2wpkhOutput(publicKeyHash)).toMatchSnapshot();
  });

  test.each([
    [OutputType.Bech32, p2wshOutput],
    [OutputType.Compatibility, p2shP2wshOutput],
    [OutputType.Legacy, p2shOutput],
    [OutputType.Taproot, undefined],
    [42 as OutputType, undefined],
  ])(
    'should get correct SH function for output type %s',
    (type: OutputType, expectedFunc: typeof p2shOutput | undefined) => {
      expect(outputFunctionForType(type)).toEqual(expectedFunc);
    },
  );
});
