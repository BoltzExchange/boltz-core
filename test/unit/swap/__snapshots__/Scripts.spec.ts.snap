// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Scripts should get P2PKH output script 1`] = `"76a914000000000000000000000000000000000000000088ac"`;

exports[`Scripts should get P2SH nested P2WPKH output script 1`] = `
{
  "outputScript": {
    "data": [
      169,
      20,
      103,
      193,
      13,
      77,
      16,
      146,
      117,
      15,
      13,
      58,
      164,
      170,
      113,
      82,
      249,
      13,
      161,
      231,
      66,
      72,
      135,
    ],
    "type": "Buffer",
  },
  "redeemScript": {
    "data": [
      0,
      20,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    "type": "Buffer",
  },
}
`;

exports[`Scripts should get P2SH nested P2WSH output script 1`] = `"a91466a823e1ae9236a70fe7321f5b26b09ec422a37787"`;

exports[`Scripts should get P2SH output script 1`] = `"a9149f7fd096d37ed2c0e3f7f0cfc924beef4ffceb6887"`;

exports[`Scripts should get P2TR output script 1`] = `"512075cba84f97d7af95da496b15b34c205003aea7f0daf6b6762beec29dae5e466c"`;

exports[`Scripts should get P2WPKH output script 1`] = `"00140000000000000000000000000000000000000000"`;

exports[`Scripts should get P2WSH output script 1`] = `"00206e340b9cffb37a989ca544e6bb780a2c78901d3fb33738768511a30617afa01d"`;
